components/script/dom/bluetooth.rs

```diff
@@ -40,59 +57,189 @@ impl Bluetooth {
+         filter.namePrefix.is_some()) {
+        return Err(Type(FILTER_ERROR.to_owned()));
+    }
+
+    let mut services: Vec<String> = vec!();
+    if filter.services.is_some() {
+        if filter.services.clone().unwrap().is_empty() {
+            return Err(Type(SERVICE_ERROR.to_owned()));
+        }
+        for service in filter.services.clone().unwrap() {
+            let validated_service = BluetoothUUID::GetService(global, service);
+            match validated_service {
+                Ok(valid) => services.push(valid.to_string()),
+                Err(err) => return Err(err),
+            }
+        }
+    }
+
+    let mut name = String::new();
+    if filter.name.is_some() {
+        // DOMString::len() gives back the size in bytes
+        let name_length = filter.name.clone().unwrap().len();
+        if name_length > MAX_DEVICE_NAME_LENGTH {
+            return Err(Type(NAME_TOO_LONG_ERROR.to_owned()));
+        }
+        if name_length > MAX_FILTER_NAME_LENGTH {
+            return Err(Type(FILTER_NAME_TOO_LONG_ERROR.to_owned()));
+        }
+        name = filter.name.clone().unwrap_or(DOMString::from("")).to_string();
+    }
+
+    let mut name_prefix = String::new();
+    if filter.namePrefix.is_some() {
+        let nameprefix_length = filter.namePrefix.clone().unwrap().len();
+        if nameprefix_length == 0 {
+            return Err(Type(NAME_PREFIX_ERROR.to_owned()));
+        }
+        if nameprefix_length > MAX_DEVICE_NAME_LENGTH {
+            return Err(Type(NAME_TOO_LONG_ERROR.to_owned()));
+        }
+        if nameprefix_length > MAX_FILTER_NAME_LENGTH {
+            return Err(Type(FILTER_NAME_TOO_LONG_ERROR.to_owned()));
+        }
+        name_prefix = filter.namePrefix.clone().unwrap_or(DOMString::from("")).to_string();
+    }
+
+    Ok(BluetoothScanfilter::new(name, name_prefix, services))
+}
+
+fn convert_request_device_options(options: &RequestDeviceOptions, global: GlobalRef)
+    -> Fallible<RequestDeviceoptions> {
+    if options.filters.is_empty() {
+        return Err(Type(FILTER_EMPTY_ERROR.to_owned()));
+    }
+
+    let options = options.clone();
+    let mut filters = vec!();
+    for filter in options.filters {
+        match canonicalize_filter(&filter, global) {
+            Ok(canonicalized_filter) => filters.push(canonicalized_filter),
+            Err(err) => return Err(err),
+        }
+    }
+
+    let mut optional_services = vec!();
+    if let Some(opt_services) = options.optionalServices {
+        for opt_service in opt_services {
+            match BluetoothUUID::GetService(global, opt_service) {
+                Ok(valid_service) => optional_services.push(valid_service.to_string()),
+                Err(err) => return Err(err),
+            }
+        }
+    }
+    let filters = BluetoothScanfilterSequence::new(filters);
+    let optional_services = ServiceUUIDSequence::new(optional_services);
+
+    Ok(RequestDeviceoptions::new(filters, optional_services))
```

Comment:
maybe
`RequestDeviceoptions::new(BluetoothScanfilterSequence::new(filters),
                                               ServiceUUIDSequence::new(optional_services))`
i don't think we need to put them into vars first, but with this borrow system i not so sure, and i did't tested it, so let me know what you think.